File: apps/api/app/assessment_targets.py
"""The lowest level a client may TARGET, for each self-assessment ladder.

## One product rule, two ladders

Level 1 is where an organization starts, not something it sets out to reach,
so neither target picker offers it. That is a single product decision; it is
expressed twice only because ZT counts in CISA/DoD *stages* and CSF counts in
*tiers*. They are declared side by side, under one docstring, so that the next
person to change one is looking at the other.

## This is a COPY of `apps/web/src/lib/assessment-targets.ts`

Duplicated across the LANGUAGE boundary rather than derived, because there is
no build step shared by this app and the web bundle -- the same conclusion
`SCHEMA_REASON_PREFIX` reached one module over, for the same reason.

**What closes the window, stated exactly rather than implied.** Each side's own
suite spells the number and names the other file in its failure message:
`tests/unit/test_intake_target_floor.py` here, and
`lib/intake/target-options-are-derived.test.ts` there. So a unilateral change
goes RED on the side that made it, pointing at the side that did not.

That is all it does. It does **not** prove the two agree, and nothing available
can: the api container mounts `./apps/api` alone and the web container mounts
`apps/web`, `packages`, `package.json`, `pnpm-workspace.yaml` and the lockfile
-- neither can read the other's file, so a parity check would pass in CI and
fail on every developer's machine. An author who edits the number here, updates
the assertion here, and stops has not been stopped. A shared parity gate needs
a mount that does not exist today; tracked in **#422**.

## Why the ranges are NOT declared here

A floor is one end. The other end is the ladder's length, and the ladder is
framework-specific: `csf/maturity.py::TIER_DEFINITIONS` has four entries and
`zt/maturity.py::level_count` returns 4 for CISA and 3 for DoD. Restating
either here would be a third spelling of a number those modules already own.
`routes/intake.py::_validate_targets` is where the two ends meet, because it
is the only place that can see `service_type`.

## Why the floor is NOT a `Field(ge=2)` bound

It was, on four fields, and that is #406. A declarative bound is refused by
FastAPI's own handler as `"Request validation failed."` under a `schema_*`
reason with no client copy behind it -- `CLAUDE.md` records the shape, and
`describe-save-error.ts::serverReason` withholds every `schema_*` code by
design, so the client's intake wizard rendered a bare generic fallback naming
neither the field nor the range. Moving the check into `_validate_targets`
buys a typed `{reason, message}` refusal (the D-016 pattern) on the surface
where a client first chooses a target.
"""

from __future__ import annotations

#: ZT (CISA ZTMM 2.0 and DoD ZTRA). Stage 1 is a starting point, not a goal.
MIN_TARGET_STAGE = 2

#: NIST CSF 2.0. Tier 1 ("Partial") is a starting point, not a goal.
MIN_TARGET_TIER = 2

File: apps/web/src/lib/assessment-targets.ts
/**
 * The lowest level a client may TARGET, for each self-assessment ladder.
 *
 * ## One product rule, two ladders
 *
 * Level 1 is where an organization starts, not something it sets out to
 * reach, so neither target picker offers it. That is a single product
 * decision; it is expressed twice only because ZT counts in CISA/DoD *stages*
 * and CSF counts in *tiers*. They are declared side by side, under one
 * docstring, so that the next person to change one is looking at the other.
 *
 * ## Why these are HERE and not next to the pickers
 *
 * `MIN_TARGET_STAGE` was declared in `components/admin/zt/ZtWorkspace.tsx`
 * and honoured there, while three sibling filters spelled the same floor as a
 * bare `2` (#194):
 *
 *     ZtGapList.tsx          .filter((s) => s.stage >= 2)
 *     ZtSelfAssessment.tsx   catalog.stages.filter((s) => s.stage >= 2)
 *     CsfSelfAssessment.tsx  catalog.tiers.filter((t) => t.tier >= 2)
 *
 * **All four agreed, so there was no live defect** -- and the hazard was not
 * the disagreement, it was that nothing could ever cause one to be noticed.
 * Raise the floor at the picker without the constant and `normalizeTarget`
 * can return a stage the dropdown no longer contains: a controlled `<select>`
 * whose `selectedIndex` is `-1` renders BLANK, with no error, no refusal and
 * nothing in a log. The failure is silent by construction.
 *
 * A client surface must not import from an admin component either, so
 * `ZtSelfAssessment` and `CsfSelfAssessment` had no honest way to reach the
 * existing constant. This module is the shared home that gives them one.
 *
 * ## This floor is the WEB's, not the API's
 *
 * Stated because the obvious assumption is wrong and would be load-bearing.
 * `routes/zt.py` validates a target only against the framework's own ladder
 * -- `{framework} has stages 1-{max}; target_stage={n} is not one of them.`
 * -- so the API ACCEPTS stage 1, and `app/zt/scoring.py` exports
 * `DEFAULT_TARGET_STAGE` but no minimum at all. Nothing on the Python side
 * mirrors these values, so there is no cross-language window to keep closed
 * and no server-side test that could go red if they changed.
 *
 * That cuts both ways, and it is the reason this file exists rather than a
 * comment: the web layer is the ONLY place this rule is written down, so a
 * single home is the whole of its enforcement.
 */

/** ZT (CISA ZTMM 2.0 and DoD ZTRA). Stage 1 is a starting point, not a goal. */
export const MIN_TARGET_STAGE = 2;

/** NIST CSF 2.0. Tier 1 ("Partial") is a starting point, not a goal. */
export const MIN_TARGET_TIER = 2;
